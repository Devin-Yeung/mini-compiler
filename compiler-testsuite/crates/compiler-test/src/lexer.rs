#[cfg(test)]
mod tests {
    use mini_compiler_rs::lexer::Lexer;
    use testsuite::unittest;

    unittest!(example_1, |_, src| {
        let token = Lexer::new(src).next_token();
        insta::assert_debug_snapshot!(token);
    });
}
