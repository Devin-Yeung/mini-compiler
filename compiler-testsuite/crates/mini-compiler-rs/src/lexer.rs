use std::ffi::CString;

use crate::gen::{self, destroy_lexer, lexer_next_token};
use crate::token::{Token, TokenTy};

pub struct Lexer {
    inner: *mut gen::Lexer,
    _src: CString,
}

impl Lexer {
    pub fn new(src: &str) -> Lexer {
        let c_string = CString::new(src).unwrap();
        let inner = unsafe { gen::lexer_new(c_string.as_ptr()) };

        Lexer {
            inner,
            _src: c_string,
        }
    }

    pub fn next_token(&mut self) -> Token {
        let token: Token = unsafe { lexer_next_token(self.inner).into() };
        token
    }
}

impl Iterator for Lexer {
    type Item = Token;

    fn next(&mut self) -> Option<Self::Item> {
        let token = self.next_token();
        match token.ty {
            TokenTy::Eof => None,
            _ => Some(token),
        }
    }
}

impl Drop for Lexer {
    fn drop(&mut self) {
        unsafe {
            destroy_lexer(self.inner);
        }
    }
}

#[cfg(test)]
mod tests {
    use super::Lexer;

    #[test]
    fn it_works() {
        let tokens = Lexer::new("abc").into_iter().collect::<Vec<_>>();
        assert_eq!(tokens.len(), 1);
    }
}
