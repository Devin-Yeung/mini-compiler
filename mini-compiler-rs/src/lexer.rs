use std::ffi::CString;

use crate::gen::{self, destroy_lexer, lexer_next_token};
use crate::token::Token;

pub struct Lexer {
    inner: *mut gen::Lexer,
    _src: CString,
}

impl Lexer {
    pub fn new(src: &str) -> Lexer {
        let c_string = CString::new(src).unwrap();
        let inner = unsafe { gen::lexer_new(c_string.as_ptr()) };

        Lexer {
            inner,
            _src: c_string,
        }
    }

    pub fn next_token(&mut self) -> Token {
        let token: Token = unsafe { lexer_next_token(self.inner).into() };
        token
    }
}

impl Drop for Lexer {
    fn drop(&mut self) {
        unsafe {
            destroy_lexer(self.inner);
        }
    }
}

#[cfg(test)]
mod tests {
    use super::Lexer;

    #[test]
    fn it_works() {
        let mut lexer = Lexer::new("abc");
        println!("{:?}", lexer.next_token());
        println!("{:?}", lexer.next_token());
        println!("{:?}", lexer.next_token());
        println!("{:?}", lexer.next_token());
    }
}
