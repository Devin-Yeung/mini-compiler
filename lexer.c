#include "lexer.h"

#include <ctype.h>
#include <stdbool.h>
#include <stdint.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#ifdef LOG
#include "log.h"
#else
// If log is not enable, just ignore the macro call
#define log_trace(...)
#define log_debug(...)
#define log_info(...)
#define log_warn(...)
#define log_error(...)
#define log_fatal(...)
#endif

void dfa_reset(DFA *dfa) { dfa->state = dfa->start; };

void dfa_next(DFA *dfa, char c) {
    dfa->state = dfa->table[(int)c][dfa->state];
};

bool dfa_is_accept(const DFA *dfa) {
    bool accept = false;
    for (unsigned i = 0; i < dfa->n_accepts; i++) {
        accept |= (dfa->state == dfa->accepts[i]);
    }
    return accept;
};

bool dfa_matches(DFA *dfa, const char *s) {
    unsigned cursor = 0;
    while (s[cursor] != '\0') {
        dfa_next(dfa, s[cursor++]);
    }

    bool accept = dfa_is_accept(dfa);
    dfa_reset(dfa);
    return accept;
};

Token *eof_token() {
    Token *token = (Token *)malloc(sizeof(Token));
    token->lexeme = (char *)malloc(sizeof(char));
    token->lexeme[0] = '\0';
    token->span = NULL;
    token->ty = Eof;
    return token;
}

char *debug_token(Token *token) {
    size_t bufsz = 256;
    char *buf = (char *)malloc(sizeof(char) * bufsz);
    switch (token->ty) {
        case BoolLit: /* Boolean Literal */
            snprintf(buf, bufsz, "BoolLit(%s) at Span (%u, %u)", token->lexeme,
                     token->span->start, token->span->end);
            break;
        case NatLit: /* Natural Number Literal */
            snprintf(buf, bufsz, "NatLit(%s)  at Span (%u, %u)", token->lexeme,
                     token->span->start, token->span->end);
            break;
        case BoolDecl: /* The variable type Boolean */
            snprintf(buf, bufsz, "BoolDecl    at Span (%u, %u)",
                     token->span->start, token->span->end);
            break;
        case NatDecl: /* The variable type Natural */
            snprintf(buf, bufsz, "NatDecl     at Span (%u, %u)",
                     token->span->start, token->span->end);
            break;
        case FuncDecl: /* The variable type Function */
            snprintf(buf, bufsz, "FuncDecl    at Span (%u, %u)",
                     token->span->start, token->span->end);
            break;
        case Comma: /* , */
            snprintf(buf, bufsz, "Token(',')  at Span (%u, %u)",
                     token->span->start, token->span->end);
            break;
        case QuestionMark: /* ? */
            snprintf(buf, bufsz, "Token('?')  at Span (%u, %u)",
                     token->span->start, token->span->end);
            break;
        case Colon: /* : */
            snprintf(buf, bufsz, "Token(':')  at Span (%u, %u)",
                     token->span->start, token->span->end);
            break;
        case Semicolon: /* ; */
            snprintf(buf, bufsz, "Token(';')  at Span (%u, %u)",
                     token->span->start, token->span->end);
            break;
        case LeftParen: /* ( */
            snprintf(buf, bufsz, "Token('(')  at Span (%u, %u)",
                     token->span->start, token->span->end);
            break;
        case RightParen: /* ) */
            snprintf(buf, bufsz, "Token(')')  at Span (%u, %u)",
                     token->span->start, token->span->end);
            break;
        case Plus: /* + */
            snprintf(buf, bufsz, "Token('+')  at Span (%u, %u)",
                     token->span->start, token->span->end);
            break;
        case Ampersand: /* & */
            snprintf(buf, bufsz, "Token('&')  at Span (%u, %u)",
                     token->span->start, token->span->end);
            break;
        case Arrow: /* -> */
            snprintf(buf, bufsz, "Token('->') at Span (%u, %u)",
                     token->span->start, token->span->end);
            break;
        case Less: /* < */
            snprintf(buf, bufsz, "Token('<')  at Span (%u, %u)",
                     token->span->start, token->span->end);
            break;
        case Equal: /* = */
            snprintf(buf, bufsz, "Token('=')  at Span (%u, %u)",
                     token->span->start, token->span->end);
            break;
        case Comment: /* [] */
            snprintf(buf, bufsz, "Comment(%s) at Span (%u, %u)", token->lexeme,
                     token->span->start, token->span->end);
            break;
        case Identifier: /* Identifier */
            snprintf(buf, bufsz, "Ident(%s)   at Span (%u, %u)", token->lexeme,
                     token->span->start, token->span->end);
            break;
        case Eof: /* End of file */
            snprintf(buf, bufsz, "EOF\n");
            break;
        case Invalid: /* Invalid token */
            snprintf(buf, bufsz, "Invalid(%s) at Span (%u, %u)", token->lexeme,
                     token->span->start, token->span->end);
            break;
    }
    return buf;
}

Span *span_new(unsigned start, unsigned end) {
    Span *span = (Span *)malloc(sizeof(Span));
    span->start = start;
    span->end = end;
    return span;
}

void destroy_span(Span *span) {
    if (span == NULL) {
        return;
    }
    free(span);
}

TokenTy get_token_type(char *lexeme) {
    switch (lexeme[0]) {
        case ',':
            return Comma;
        case '?':
            return QuestionMark;
        case ':':
            return Colon;
        case ';':
            return Semicolon;
        case '(':
            return LeftParen;
        case ')':
            return RightParen;
        case '+':
            return Plus;
        case '&':
            return Ampersand;
        case '-': /* -> is the only valid token start with - sign */
            return Arrow;
        case '<':
            return Less;
        case '=':
            return Equal;
        case '[':
            return Comment;
    }

    if (strcmp(lexeme, "fun") == 0) {
        return FuncDecl;
    }

    if (strcmp(lexeme, "T") == 0 || strcmp(lexeme, "F") == 0) {
        return BoolLit;
    }

    if (strcmp(lexeme, "bool") == 0) {
        return BoolDecl;
    }

    if (strcmp(lexeme, "nat") == 0) {
        return NatDecl;
    }

    if (isdigit(lexeme[0])) {
        return NatLit;
    } else {
        return Identifier;
    }
}

bool need_to_check_dfa(char cur, char peek, bool comment) {
    // never check dfa if not reach the end of comment
    if (cur != ']' && comment) {
        return false;
    }

    if (isalpha(cur)) {
        return !isalpha(peek) && !isdigit(peek);
    }
    if (isdigit(cur)) {
        return !isalpha(peek) && !isdigit(peek);
    }

    if (cur == '-' && peek == '>') {
        return false;
    }
    return true;
}

char *pretty_ascii(char ascii, char *buf, size_t bufsz) {
    if (ascii >= '!') {
        snprintf(buf, bufsz, "%c", ascii);
    } else if (isspace(ascii) || ascii == '\0') {
        switch (ascii) {
            case 0x0:
                snprintf(buf, bufsz, "\\0");  // NUL char
            case 0x20:
                snprintf(buf, bufsz, "space");  // Space
                break;
            case 0x0c:
                snprintf(buf, bufsz, "\\f");  // Form feed
                break;
            case 0x0a:
                snprintf(buf, bufsz, "\\n");  // Line feed
                break;
            case 0x0d:
                snprintf(buf, bufsz, "\\r");  // Carriage return
                break;
            case 0x09:
                snprintf(buf, bufsz, "\\t");  // Horizontal tab
                break;
            case 0x0b:
                snprintf(buf, bufsz, "\\v");  // Vertical tab
                break;
        }
    } else {
        snprintf(buf, bufsz, "%u", (uint8_t)ascii);
    }
    return buf;
}

Lexer *lexer_new(const char *src) {
    Lexer *lexer = malloc(sizeof(Lexer));

    lexer->dfa = &LEXER_DFA;
    lexer->start = 0;
    lexer->end = 0;
    lexer->src = src;
    lexer->len = strlen(src);

    return lexer;
}

Token *lexer_next_token(Lexer *lexer) {
    log_debug("Getting next token, start from %u", lexer->start);

    if (lexer->start > lexer->len) {
        log_info("EOF reached");
        return eof_token();
    }

    unsigned cursor = lexer->start;
    bool comment = false;
    while (cursor <= lexer->len) {
        switch (lexer->src[cursor]) {
            case '[':
                comment = true;
                break;
            case ']':
                comment = false;
                log_debug("Comment End");
                break;
        }
        char buf[8];
        log_debug("Cursor at src[%u] = [%s]", cursor,
                  pretty_ascii(lexer->src[cursor], buf, 8));
        char peek = cursor + 1 < lexer->len ? lexer->src[cursor + 1] : '\0';
        log_debug("Peek   at src[%u] = [%s]", cursor + 1,
                  pretty_ascii(peek, buf, 8));
        // Don't skip whitespace in comment section
        if (!comment &&
            (isspace(lexer->src[cursor]) || lexer->src[cursor] == '\0')) {
            log_debug("Whitespace detected at %u, Skip", cursor);
            cursor += 1;
            // reset span
            lexer->start = cursor;
            lexer->end = cursor;
        } else {
            // feed the current char to dfa
            dfa_next(lexer->dfa, lexer->src[cursor]);
            // if next char is whitespace or '\0'
            // which indicates current token is done
            if (need_to_check_dfa(lexer->src[cursor], peek, comment)) {
                log_debug(
                    "DFA checking condition satisfied, check DFA to decide "
                    "whether to accept");

                Token *token = (Token *)malloc(sizeof(Token));
                unsigned token_len = lexer->end - lexer->start + 1;
                char *lexeme = (char *)malloc(sizeof(char) * (token_len + 1));
                token->lexeme =
                    strncpy(lexeme, lexer->src + lexer->start, token_len);
                token->lexeme[token_len] = '\0';
                token->span = span_new(lexer->start, lexer->end);

                if (dfa_is_accept(lexer->dfa)) {
                    token->ty = get_token_type(token->lexeme);
                    log_debug("accept:「%s」 => span(%u, %u)", token->lexeme,
                              lexer->start, lexer->end);
                } else {
                    token->ty = Invalid;
                    log_warn("The lexeme is rejected by the DFA");
                }
                dfa_reset(lexer->dfa);
                lexer->start = lexer->end + 1;
                lexer->end = lexer->start;
                return token;
            } else {
                lexer->end += 1;
            }

            cursor += 1;
        }
    }

    return eof_token();
}

void destroy_token(Token *token) {
    destroy_span(token->span);
    free(token->lexeme);
    free(token);
}

void destroy_lexer(Lexer *lexer) { free(lexer); }
