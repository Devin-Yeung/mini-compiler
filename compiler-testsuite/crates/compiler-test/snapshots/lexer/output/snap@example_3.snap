---
source: crates/compiler-test/src/lexer.rs
expression: token
---
[
    Token {
        lexeme: "fun",
        ty: FuncDecl,
        span: Span {
            start: 0,
            end: 2,
        },
    },
    Token {
        lexeme: "recp",
        ty: Identifier,
        span: Span {
            start: 4,
            end: 7,
        },
    },
    Token {
        lexeme: "nat",
        ty: NatDecl,
        span: Span {
            start: 9,
            end: 11,
        },
    },
    Token {
        lexeme: "x",
        ty: Identifier,
        span: Span {
            start: 13,
            end: 13,
        },
    },
    Token {
        lexeme: "nat",
        ty: NatDecl,
        span: Span {
            start: 15,
            end: 17,
        },
    },
    Token {
        lexeme: "i",
        ty: Identifier,
        span: Span {
            start: 19,
            end: 19,
        },
    },
    Token {
        lexeme: "nat",
        ty: NatDecl,
        span: Span {
            start: 21,
            end: 23,
        },
    },
    Token {
        lexeme: "j",
        ty: Identifier,
        span: Span {
            start: 25,
            end: 25,
        },
    },
    Token {
        lexeme: "->",
        ty: Arrow,
        span: Span {
            start: 27,
            end: 28,
        },
    },
    Token {
        lexeme: "x",
        ty: Identifier,
        span: Span {
            start: 31,
            end: 31,
        },
    },
    Token {
        lexeme: "=",
        ty: Equal,
        span: Span {
            start: 33,
            end: 33,
        },
    },
    Token {
        lexeme: "0",
        ty: Literal,
        span: Span {
            start: 35,
            end: 35,
        },
    },
    Token {
        lexeme: "&",
        ty: Ampersand,
        span: Span {
            start: 37,
            end: 37,
        },
    },
    Token {
        lexeme: "i",
        ty: Identifier,
        span: Span {
            start: 39,
            end: 39,
        },
    },
    Token {
        lexeme: "<",
        ty: Less,
        span: Span {
            start: 41,
            end: 41,
        },
    },
    Token {
        lexeme: "x",
        ty: Identifier,
        span: Span {
            start: 43,
            end: 43,
        },
    },
    Token {
        lexeme: "?",
        ty: QuestionMark,
        span: Span {
            start: 45,
            end: 45,
        },
    },
    Token {
        lexeme: "0",
        ty: Literal,
        span: Span {
            start: 47,
            end: 47,
        },
    },
    Token {
        lexeme: ":",
        ty: Colon,
        span: Span {
            start: 49,
            end: 49,
        },
    },
    Token {
        lexeme: "0",
        ty: Literal,
        span: Span {
            start: 52,
            end: 52,
        },
    },
    Token {
        lexeme: "<",
        ty: Less,
        span: Span {
            start: 54,
            end: 54,
        },
    },
    Token {
        lexeme: "x",
        ty: Identifier,
        span: Span {
            start: 56,
            end: 56,
        },
    },
    Token {
        lexeme: "&",
        ty: Ampersand,
        span: Span {
            start: 58,
            end: 58,
        },
    },
    Token {
        lexeme: "x",
        ty: Identifier,
        span: Span {
            start: 60,
            end: 60,
        },
    },
    Token {
        lexeme: "=",
        ty: Equal,
        span: Span {
            start: 62,
            end: 62,
        },
    },
    Token {
        lexeme: "i",
        ty: Identifier,
        span: Span {
            start: 64,
            end: 64,
        },
    },
    Token {
        lexeme: "?",
        ty: QuestionMark,
        span: Span {
            start: 66,
            end: 66,
        },
    },
    Token {
        lexeme: "j",
        ty: Identifier,
        span: Span {
            start: 68,
            end: 68,
        },
    },
    Token {
        lexeme: ":",
        ty: Colon,
        span: Span {
            start: 70,
            end: 70,
        },
    },
    Token {
        lexeme: "(",
        ty: LeftParen,
        span: Span {
            start: 73,
            end: 73,
        },
    },
    Token {
        lexeme: "recp",
        ty: Identifier,
        span: Span {
            start: 74,
            end: 77,
        },
    },
    Token {
        lexeme: "x",
        ty: Identifier,
        span: Span {
            start: 79,
            end: 79,
        },
    },
    Token {
        lexeme: "i",
        ty: Identifier,
        span: Span {
            start: 81,
            end: 81,
        },
    },
    Token {
        lexeme: "+",
        ty: Plus,
        span: Span {
            start: 82,
            end: 82,
        },
    },
    Token {
        lexeme: "1",
        ty: Literal,
        span: Span {
            start: 83,
            end: 83,
        },
    },
    Token {
        lexeme: "j",
        ty: Identifier,
        span: Span {
            start: 85,
            end: 85,
        },
    },
    Token {
        lexeme: "+",
        ty: Plus,
        span: Span {
            start: 86,
            end: 86,
        },
    },
    Token {
        lexeme: "1",
        ty: Literal,
        span: Span {
            start: 87,
            end: 87,
        },
    },
    Token {
        lexeme: ")",
        ty: RightParen,
        span: Span {
            start: 88,
            end: 88,
        },
    },
    Token {
        lexeme: ";",
        ty: Semicolon,
        span: Span {
            start: 90,
            end: 90,
        },
    },
    Token {
        lexeme: "fun",
        ty: FuncDecl,
        span: Span {
            start: 92,
            end: 94,
        },
    },
    Token {
        lexeme: "pred",
        ty: Identifier,
        span: Span {
            start: 96,
            end: 99,
        },
    },
    Token {
        lexeme: "nat",
        ty: NatDecl,
        span: Span {
            start: 101,
            end: 103,
        },
    },
    Token {
        lexeme: "x",
        ty: Identifier,
        span: Span {
            start: 105,
            end: 105,
        },
    },
    Token {
        lexeme: "->",
        ty: Arrow,
        span: Span {
            start: 107,
            end: 108,
        },
    },
    Token {
        lexeme: "T",
        ty: Literal,
        span: Span {
            start: 110,
            end: 110,
        },
    },
    Token {
        lexeme: "?",
        ty: QuestionMark,
        span: Span {
            start: 112,
            end: 112,
        },
    },
    Token {
        lexeme: "(",
        ty: LeftParen,
        span: Span {
            start: 114,
            end: 114,
        },
    },
    Token {
        lexeme: "recp",
        ty: Identifier,
        span: Span {
            start: 115,
            end: 118,
        },
    },
    Token {
        lexeme: "x",
        ty: Identifier,
        span: Span {
            start: 120,
            end: 120,
        },
    },
    Token {
        lexeme: "1",
        ty: Literal,
        span: Span {
            start: 122,
            end: 122,
        },
    },
    Token {
        lexeme: "0",
        ty: Literal,
        span: Span {
            start: 124,
            end: 124,
        },
    },
    Token {
        lexeme: ")",
        ty: RightParen,
        span: Span {
            start: 125,
            end: 125,
        },
    },
    Token {
        lexeme: ":",
        ty: Colon,
        span: Span {
            start: 127,
            end: 127,
        },
    },
    Token {
        lexeme: "0",
        ty: Literal,
        span: Span {
            start: 129,
            end: 129,
        },
    },
    Token {
        lexeme: ";",
        ty: Semicolon,
        span: Span {
            start: 130,
            end: 130,
        },
    },
    Token {
        lexeme: "T",
        ty: Literal,
        span: Span {
            start: 132,
            end: 132,
        },
    },
    Token {
        lexeme: "?",
        ty: QuestionMark,
        span: Span {
            start: 134,
            end: 134,
        },
    },
    Token {
        lexeme: "(",
        ty: LeftParen,
        span: Span {
            start: 136,
            end: 136,
        },
    },
    Token {
        lexeme: "pred",
        ty: Identifier,
        span: Span {
            start: 137,
            end: 140,
        },
    },
    Token {
        lexeme: "2",
        ty: Literal,
        span: Span {
            start: 142,
            end: 142,
        },
    },
    Token {
        lexeme: ")",
        ty: RightParen,
        span: Span {
            start: 143,
            end: 143,
        },
    },
    Token {
        lexeme: ":",
        ty: Colon,
        span: Span {
            start: 145,
            end: 145,
        },
    },
    Token {
        lexeme: "0",
        ty: Literal,
        span: Span {
            start: 147,
            end: 147,
        },
    },
]
